### PR TITLE
Add socket_connect_timeout option.

### PR DESCRIPTION
**Pull request by benjie**
This allows you to specify the connection timeout for the stream; note that once the stream is connected the existing code performs `stream.setTimeout(0)` thereby preventing this patch triggering an idle timeout on an established connection. (This is also why I called it `socket_connect_timeout` rather than simply `socket_timeout`.)

I added `this.stream.setTimeout` to `install_stream_listeners` as it seemed the DRY-est place to put it since it is called for all streams just after they're created. If this is not the right place, please let me know.

### CODE REVIEW

**Comment by BridgeAR** [comment]
@bengl LGTM would you please rebase and add some tests?

**Comment by bengl** [comment]
by @bengl you mean @benjie right? :smile:

**Comment by BridgeAR** [comment]
@bengl ups sorry for bringing you up :D 
And yes, I meant @benjie

**Comment by benjie** [comment]
:laughing: thanks @bengl :)

@BridgeAR I'm not sure how I'd write a test for this

**Comment by benjie** [comment]
Rebased :)

File: index.js [hunk 1]
```diff
@@ -218,6 +224,20 @@ RedisClient.prototype.do_auth = function () {
...
 };
 
+RedisClient.prototype.on_timeout = function () {
+    if (this.connected || this.closing) {
+        // Only handle connection timeouts.
+        return;
+    }
+
+    if (exports.debug_mode) {
+        console.log("Stream timeout " + this.address + " id " + this.connection_id);
```
Comment (BridgeAR) [inline index.js]: Please use
```js
debug("Stream timeout " + this.address + " id " + this.connection_id);
```
instead

**Comment by BridgeAR** [comment]
@benjie shouldn't it be enough if you just manipulate the socket timeout with the parameter? What you want to reach is to have a individual timeout instead of the system one, so everything else should stay the same?

And a test could be written by checking the system timeout (use a before, create a client that can't connect and check for the timeout and safe that value). Now you can increase or decrease that value as you want and do the same again and check that the timeout triggered at another time.

**Comment by benjie** [comment]
What endpoint should I use that it times out connecting to? I think a closed port won't do because it'll just drop the connection immediately (no chance for a timeout) so it'd need to be one that drops the packets. Adding an `iptables` filter would make the test platform dependent; is there a cross-platform `/dev/null` equivalent for TCP?

**Comment by benjie** [comment]
(`debug()` change implemented)

**Comment by BridgeAR** [comment]
You could mock the stream :)

Google just gave me this: https://gist.github.com/mjackson/1201196

**Comment by BridgeAR** [comment]
@benjie I used your work to use the connection_timeout as socket timeout if explicitly provided and mentioned you in the changelog. That way we do not have to add another option and most people thought the connection timeout would also set a socket timeout.

Thx a lot for your help!

**Comment by benjie** [comment]
Thanks @BridgeAR; sorry I didn't get around to writing the tests yet but your solution's probably better anyway! :smile:

**Comment by benjie** [comment]
Hey @BridgeAR; I think your fix serves a different purpose to my PR - I want the connection attempt to fail, be closed, and then attempt again in a much tighter loop than the OS default (like every 30 seconds rather than 180); whereas `connect_timeout` stops attempting to reconnect meaning we remain disconnected forever. There may still be value in this PR

**Comment by billinghamj** [comment]
This should really be reopened

**Comment by BridgeAR** [comment]
@benjie thanks for pointing this out and sorry for not understanding your original issue properly right away.

There seems to be a new PR to do the same: https://github.com/NodeRedis/node_redis/pull/1430

**Comment by benjie** [comment]
No problem :)
